bus-mapping/src/evm/opcodes/codecopy.rs

```diff
@@ -0,0 +1,238 @@
+
+use super::Opcode;
+
+#[derive(Clone, Copy, Debug)]
+pub(crate) struct Codecopy;
+
+impl Opcode for Codecopy {
+    fn gen_associated_ops(
+        state: &mut CircuitInputStateRef,
+        geth_steps: &[GethExecStep],
+    ) -> Result<Vec<ExecStep>, Error> {
+        let geth_step = &geth_steps[0];
+        let mut exec_steps = vec![gen_codecopy_step(state, geth_step)?];
+        let memory_copy_steps = gen_memory_copy_steps(state, geth_steps)?;
+        exec_steps.extend(memory_copy_steps);
+        Ok(exec_steps)
+    }
+}
+
+fn gen_codecopy_step(
+    state: &mut CircuitInputStateRef,
+    geth_step: &GethExecStep,
+) -> Result<ExecStep, Error> {
+    let mut exec_step = state.new_step(geth_step)?;
+
+    let memory_offset = geth_step.stack.nth_last(0)?;
+    let code_offset = geth_step.stack.nth_last(1)?;
+    let length = geth_step.stack.nth_last(2)?;
+
+    // stack reads
+    state.push_stack_op(
+        &mut exec_step,
+        RW::READ,
+        geth_step.stack.nth_last_filled(0),
+        memory_offset,
+    )?;
+    state.push_stack_op(
+        &mut exec_step,
+        RW::READ,
+        geth_step.stack.nth_last_filled(1),
+        code_offset,
+    )?;
+    state.push_stack_op(
+        &mut exec_step,
+        RW::READ,
+        geth_step.stack.nth_last_filled(2),
+        length,
+    )?;
+    Ok(exec_step)
+}
+
+fn gen_memory_copy_step(
+    state: &mut CircuitInputStateRef,
+    exec_step: &mut ExecStep,
+    aux_data: CopyCodeToMemoryAuxData,
+    code: &eth_types::Bytecode,
+) -> Result<(), Error> {
+    let CopyCodeToMemoryAuxData {
+        src_addr,
+        dst_addr,
+        bytes_left,
+        src_addr_end,
+        code_source,
+    } = aux_data;
+
+    for idx in 0..std::cmp::min(bytes_left as usize, MAX_COPY_BYTES) {
+        let addr = (src_addr as usize) + idx;
+        let byte = if addr < (src_addr_end as usize) {
+            code.code()[addr]
+        } else {
+            0
+        };
+        state.push_memory_op(
+            exec_step,
+            RW::WRITE,
+            ((dst_addr as usize) + idx).into(),
+            byte,
+        )?;
+    }
+
+    exec_step.aux_data = Some(StepAuxiliaryData::CopyCodeToMemory(
+        CopyCodeToMemoryAuxData {
+            src_addr: src_addr as u64,
+            dst_addr: dst_addr as u64,
+            bytes_left: bytes_left as u64,
+            src_addr_end: src_addr_end as u64,
+            code_source,
+        },
+    ));
+
+    Ok(())
+}
+
+fn gen_memory_copy_steps(
+    state: &mut CircuitInputStateRef,
+    geth_steps: &[GethExecStep],
+) -> Result<Vec<ExecStep>, Error> {
+    let memory_offset = geth_steps[0].stack.nth_last(0)?.as_u64();
+    let code_offset = geth_steps[0].stack.nth_last(1)?.as_u64();
+    let length = geth_steps[0].stack.nth_last(2)?.as_u64();
+
+    let code_source = state.call()?.code_hash;
+    let code = state.code_db.0.get(&code_source).unwrap();
+    let code = Bytecode::try_from(code.clone()).unwrap();
```

Comment:
The `Bytecode` parsing from `Vec<u8>` looks nice, but in general case, we don't know if deployed contracts are all following our rules (enough push data, valid opcodes, etc...), or they could be data contracts, so I think this parsing could be saved.

Reply:
Got it. I think we should still keep the associated code in `opcode.rs` and `bytecode.rs` since we might need that functionality in the future. I will remove the parsing from here though.